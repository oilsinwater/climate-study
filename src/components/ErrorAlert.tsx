import { Alert } from '@mui/material';

interface ErrorAlertProps {
  message: string;
}

export const ErrorAlert = ({ message }: ErrorAlertProps) => (
  <Alert severity="error" sx={{ margin: 2 }}>
    {message}
  </Alert>
);
