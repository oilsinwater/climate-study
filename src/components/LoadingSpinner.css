.loading-spinner-container {
  display: flex;
  justify-content: center;
  padding: 20px;
}
