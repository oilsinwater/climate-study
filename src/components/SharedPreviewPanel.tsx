import React, { useState } from 'react';
import {
  Box,
  Button,
  IconButton,
  Link,
  Paper,
  Stack,
  Typography,
  Tab,
  Tabs,
} from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import { Link as RouterLink } from 'react-router-dom';
import { LabelValueTable } from './LabelValueTable';
import { DataGrid } from '@mui/x-data-grid';
import Plot from 'react-plotly.js';

interface SharedPreviewPanelProps {
  /**
   * Data for the selected item to preview
   */
  previewItem: any;
  /**
   * Function to handle closing the preview panel
   */
  onClose: () => void;
  /**
   * Optional ID field for the data items (defaults to 'id')
   */
  idField?: string;
  /**
   * Optional columns configuration for displaying the item data
   */
  columns?: Array<any>;
  /**
   * Optional configuration for detail page navigation
   */
  detailsConfig?: {
    enabled: boolean;
    path: string;
  };
}

/**
 * Shared preview panel component that can be used across different sections
 * of the application to display detailed information about a selected item.
 */
export const SharedPreviewPanel: React.FC<SharedPreviewPanelProps> = ({
  previewItem,
  onClose,
  idField = 'id',
  columns = [],
  detailsConfig = { enabled: false, path: '' },
}) => {
  const [tabValue, setTabValue] = useState(0);

  const handleTabChange = (event: React.SyntheticEvent, newValue: number) => {
    setTabValue(newValue);
  };

  // Generate visualization data based on the preview item properties
  const generateVisualizationData = () => {
    try {
      // Check for time-related or numeric fields to use in visualizations
      const numericKeys = Object.keys(previewItem || {}).filter(
        key => typeof previewItem[key] === 'number'
      );
      
      // If there are at least 2 numeric fields, create a basic visualization
      if (numericKeys.length >= 2) {
        const key1 = numericKeys[0];
        const key2 = numericKeys[1];
        
        return {
          timeSeriesData: {
            x: Array.from({ length: 20 }, (_, i) => i),
            y: Array.from({ length: 20 }, (_, i) => 
              Math.sin(i * 0.5) * previewItem[key1] / 10 + previewItem[key1]
            ),
          },
          scatterData: {
            x: [previewItem[key1], previewItem[key1] * 1.2],
            y: [previewItem[key2], previewItem[key2] * 0.8],
            text: ['Current', 'Projected'],
          }
        };
      }
      
      // Default visualization data if no suitable properties found
      return {
        timeSeriesData: {
          x: Array.from({ length: 20 }, (_, i) => i),
          y: Array.from({ length: 20 }, (_, i) => Math.sin(i * 0.5) * 5 + 10),
        },
        scatterData: {
          x: [1, 2, 3, 4],
          y: [10, 15, 13, 17],
          text: ['A', 'B', 'C', 'D'],
        }
      };
    } catch (error) {
      console.error('Error generating visualization data:', error);
      // Return safe default data
      return {
        timeSeriesData: {
          x: [0, 1, 2, 3, 4],
          y: [5, 10, 5, 10, 5],
        },
        scatterData: {
          x: [1, 2, 3, 4],
          y: [10, 15, 13, 17],
          text: ['A', 'B', 'C', 'D'],
        }
      };
    }
  };

  const visualizationData = generateVisualizationData();

  // If no item is selected, don't render anything
  if (!previewItem) {
    return null;
  }

  // Extract the item attributes to display in the details
  const attributes = columns.length > 0
    ? columns.map(col => ({
        label: col.headerName || col.field,
        value: previewItem[col.field],
        type: col.type,
        units: col.units
      }))
    : Object.entries(previewItem).map(([key, value]) => ({
        label: key,
        value,
        type: typeof value === 'number' ? 'number' : 'string'
      }));

  return (
    <Paper 
      elevation={0} 
      sx={{ 
        height: '100%', 
        display: 'flex', 
        flexDirection: 'column',
        borderLeft: '1px solid',
        borderColor: 'divider'
      }}
    >
      <Stack 
        direction="row" 
        alignItems="center" 
        justifyContent="space-between"
        sx={{ p: 2, borderBottom: '1px solid', borderColor: 'divider' }}
      >
        <Typography variant="h6" component="h2">
          {previewItem.name || previewItem.title || `Item ${previewItem[idField]}`}
        </Typography>
        <IconButton onClick={onClose} size="small">
          <CloseIcon fontSize="small" />
        </IconButton>
      </Stack>

      <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
        <Tabs 
          value={tabValue} 
          onChange={handleTabChange}
          variant="fullWidth"
        >
          <Tab label="Details" />
          <Tab label="Visualizations" />
          <Tab label="Related" />
        </Tabs>
      </Box>

      <Box sx={{ flex: 1, overflow: 'auto', p: 2 }}>
        {/* Details Tab */}
        {tabValue === 0 && (
          <Stack spacing={2}>
            <LabelValueTable rows={attributes} />
            
            {detailsConfig.enabled && (
              <Button 
                component={RouterLink} 
                to={`${detailsConfig.path}/${previewItem[idField]}`}
                variant="contained"
                fullWidth
                sx={{ mt: 2 }}
              >
                View Full Details
              </Button>
            )}
          </Stack>
        )}

        {/* Visualizations Tab */}
        {tabValue === 1 && (
          <Stack spacing={3}>
            <Box>
              <Typography variant="subtitle1" gutterBottom>
                Time Series Data
              </Typography>
              <Plot
                data={[
                  {
                    x: visualizationData.timeSeriesData.x,
                    y: visualizationData.timeSeriesData.y,
                    type: 'scatter',
                    mode: 'lines+markers',
                    marker: {color: 'blue'},
                  },
                ]}
                layout={{
                  autosize: true,
                  height: 300,
                  margin: {l: 50, r: 50, t: 30, b: 50},
                  xaxis: {title: 'Time'},
                  yaxis: {title: 'Value'},
                }}
                style={{width: '100%'}}
                useResizeHandler={true}
                onError={(error) => console.error('Error rendering Plotly chart:', error)}
              />
            </Box>
            
            <Box>
              <Typography variant="subtitle1" gutterBottom>
                Data Distribution
              </Typography>
              <Plot
                data={[
                  {
                    x: visualizationData.scatterData.x,
                    y: visualizationData.scatterData.y,
                    text: visualizationData.scatterData.text,
                    type: 'scatter',
                    mode: 'markers',
                    marker: {
                      size: 12,
                      color: 'rgba(55, 128, 191, 0.7)',
                    },
                  },
                ]}
                layout={{
                  autosize: true,
                  height: 300,
                  margin: {l: 50, r: 50, t: 30, b: 50},
                  xaxis: {title: 'X Value'},
                  yaxis: {title: 'Y Value'},
                }}
                style={{width: '100%'}}
                useResizeHandler={true}
                onError={(error) => console.error('Error rendering Plotly chart:', error)}
              />
            </Box>
          </Stack>
        )}

        {/* Related Tab */}
        {tabValue === 2 && (
          <Box>
            <Typography variant="subtitle1" gutterBottom>
              Related Items
            </Typography>
            <DataGrid
              rows={Array(5).fill(0).map((_, i) => ({
                id: i,
                name: `Related Item ${i + 1}`,
                type: 'Dataset',
                relation: 'Referenced'
              }))}
              columns={[
                { field: 'name', headerName: 'Name', flex: 1 },
                { field: 'type', headerName: 'Type', width: 120 },
                { field: 'relation', headerName: 'Relation', width: 120 }
              ]}
              pageSizeOptions={[5]}
              initialState={{
                pagination: { paginationModel: { pageSize: 5 } },
              }}
              autoHeight
              disableRowSelectionOnClick
            />
          </Box>
        )}
      </Box>
    </Paper>
  );
};
