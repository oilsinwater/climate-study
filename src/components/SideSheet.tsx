import { Drawer, List, ListItem, ListItemButton, ListItemIcon, ListItemText, IconButton } from '@mui/material';
import MenuIcon from '@mui/icons-material/Menu';
import HomeIcon from '@mui/icons-material/Home';
import SearchIcon from '@mui/icons-material/Search';
import ExploreIcon from '@mui/icons-material/Explore';
import InfoIcon from '@mui/icons-material/Info';
import { Link as RouterLink } from 'react-router-dom';
import React, { useState } from 'react';

export const SideSheet: React.FC = () => {
  const [open, setOpen] = useState(false);

  const toggleDrawer = (open: boolean) => (event: React.KeyboardEvent | React.MouseEvent) => {
    if (event.type === 'keydown' && ((event as React.KeyboardEvent).key === 'Tab' || (event as React.KeyboardEvent).key === 'Shift')) {
      return;
    }
    setOpen(open);
  };

  return (
    <>
      <IconButton onClick={toggleDrawer(true)}>
        <MenuIcon />
      </IconButton>
      <Drawer anchor='left' open={open} onClose={toggleDrawer(false)}>
        <List>
          <ListItem disablePadding>
            <ListItemButton component={RouterLink} to='/' onClick={toggleDrawer(false)}>
              <ListItemIcon><HomeIcon /></ListItemIcon>
              <ListItemText primary='Home' />
            </ListItemButton>
          </ListItem>
          <ListItem disablePadding>
            <ListItemButton component={RouterLink} to='/search-repositories' onClick={toggleDrawer(false)}>
              <ListItemIcon><SearchIcon /></ListItemIcon>
              <ListItemText primary='Search Data' />
            </ListItemButton>
          </ListItem>
          <ListItem disablePadding>
            <ListItemButton component={RouterLink} to='/explore-data' onClick={toggleDrawer(false)}>
              <ListItemIcon><ExploreIcon /></ListItemIcon>
              <ListItemText primary='Explore Data' />
            </ListItemButton>
          </ListItem>
          <ListItem disablePadding>
            <ListItemButton component={RouterLink} to='/about' onClick={toggleDrawer(false)}>
              <ListItemIcon><InfoIcon /></ListItemIcon>
              <ListItemText primary='About' />
            </ListItemButton>
          </ListItem>
        </List>
      </Drawer>
    </>
  );
};
