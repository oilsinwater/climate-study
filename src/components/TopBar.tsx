import { AppBar, Toolbar, InputBase, IconButton } from '@mui/material';
import SearchIcon from '@mui/icons-material/Search';
import React from 'react';
import { useLocation } from 'react-router-dom';

/**
 * Top navigation bar with search input and sidesheet menu
 */
export const TopBar: React.FC = () => {
  const location = useLocation();
  const showSearch = location.pathname !== '/';

  return (
    <AppBar position="static" color="primary">
      <Toolbar>
        <InputBase
          placeholder="Search..."
          sx={{ 
            ml: 2,
            flexGrow: 1,
            backgroundColor: 'rgba(255, 255, 255, 0.15)',
            borderRadius: 1,
            px: 1,
            '&:hover': {
              backgroundColor: 'rgba(255, 255, 255, 0.25)',
            },
          }}
          startAdornment={
            <SearchIcon sx={{ color: 'inherit', mr: 1 }} />
          }
        />
      </Toolbar>
    </AppBar>
  );
};
