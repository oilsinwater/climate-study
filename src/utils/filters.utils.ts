import dayjs from 'dayjs';
import { DataFilter, FilterConfig } from '../types/filters.types';

export const filterBySearchText = <T extends Record<string, unknown>>(allData: T[], searchText?: string) => {
  let filteredData = allData;
  if (searchText) {
    filteredData = allData.filter((d) => {
      const rowString = JSON.stringify(d).toLowerCase();
      return rowString.indexOf(searchText.toLowerCase()) > -1;
    });
  }
  return filteredData;
};

export const filterByDataFilters = <T extends Record<string, unknown>>(
  allData: T[],
  filters: DataFilter[] | null,
  filterConfigs: FilterConfig[] | null
) => {
  if (!filters) return allData;

  let filteredData = allData;
  if (filters.length > 0) {
    // Pre build map of filter to operator for performance boost
    const filterOperatorMap: Record<string, string | undefined> = {};
    filters.forEach((f) => {
      if (filterConfigs) {
        const filterConfig = filterConfigs.find((c) => c.field === f.field);
        filterOperatorMap[f.field] = filterConfig?.operator;
      }
    });
    filteredData = allData.filter((d) => {
      let include = true;
      // All filters have to be matched for a row to be included in the filtered data
      filters.forEach((f) => {
        let match = false;
        if (include === true) {
          switch (filterOperatorMap[f.field]) {
            case 'contains': {
              const fieldValue = d[f.field as keyof typeof d];
              if (typeof fieldValue === 'string' && typeof f.value === 'string' && fieldValue.indexOf(f.value) > -1) {
                match = true;
              }
              break;
            }
            case 'contains-one-of': {
              if (Array.isArray(f.value)) {
                f.value.forEach((v) => {
                  if (!match) {
                    const fieldValue = d[f.field as keyof typeof d];
                    if (Array.isArray(fieldValue)) {
                      if (fieldValue.indexOf(v) > -1) {
                        match = true;
                      }
                    } else if (fieldValue === v) {
                      match = true;
                    }
                  }
                });
              }
              break;
            }
            case 'equals-one-of': {
              if (Array.isArray(f.value)) {
                f.value.forEach((v) => {
                  if (!match) {
                    const fieldValue = d[f.field as keyof typeof d];
                    if (fieldValue === v) {
                      match = true;
                    }
                  }
                });
              }
              break;
            }
            case 'between-inclusive': {
              if (Array.isArray(f.value)) {
                const min = f.value[0];
                const max = f.value[1];
                const fieldValue = d[f.field as keyof typeof d];
                if (typeof fieldValue === 'number' && typeof min === 'number' && typeof max === 'number' && 
                    fieldValue >= min && fieldValue <= max) {
                  match = true;
                }
              }
              break;
            }
            case 'between-dates-inclusive': {
              const fieldValue = d[f.field as keyof typeof d];
              if (
                typeof fieldValue === 'string' &&
                Array.isArray(f.value) &&
                f.value[0] &&
                f.value[1]
              ) {
                const dateValue = dayjs(fieldValue);
                if (
                  dateValue.isAfter(f.value[0]) &&
                  dateValue.isBefore(f.value[1])
                ) {
                  match = true;
                }
              } else {
                match = true;
              }
              break;
            }
            default:
              break;
          }
        }
        if (!match) include = false;
      });
      return include;
    });
  }
  return filteredData;
};

export const filterData = <T extends Record<string, unknown>>(
  allData: T[],
  filters: DataFilter[],
  filterConfigs: FilterConfig[],
  searchText?: string
) => {
  const filteredByText = filterBySearchText(allData, searchText);
  const filteredByTextAndDataFilters = filterByDataFilters(
    filteredByText,
    filters,
    filterConfigs
  );
  return filteredByTextAndDataFilters;
};

export const initSliderTicks = (
  ticks: number | null,
  domain: number[],
  scale?: { ticks: (count: number) => number[] }
) => {
  if (ticks === 2) {
    return domain;
  } else if (ticks !== null) {
    return scale?.ticks(ticks);
  } else {
    return;
  }
};
