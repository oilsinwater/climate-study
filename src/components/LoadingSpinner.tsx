import { CircularProgress } from '@mui/material';
import './LoadingSpinner.css';

export const LoadingSpinner = () => (
  <div className="loading-spinner-container">
    <CircularProgress />
  </div>
);
