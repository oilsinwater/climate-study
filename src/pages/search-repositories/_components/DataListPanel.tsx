import FilterAltIcon from '@mui/icons-material/FilterAlt';
import SortIcon from '@mui/icons-material/Sort';
import {
  Box,
  Button,
  Pagination,
  Paper,
  Stack,
  Typography,
  ListItemText,
} from '@mui/material';
import React, { useEffect, useState, useMemo } from 'react';
import { DataListCard } from './DataListCard';
import { getListConfig, getCardFields } from '../_config/taskflow.config';
import { useListQuery } from '../../../utils/useListQuery';
import { LoadingSpinner } from '../../../components/LoadingSpinner';
import { ErrorAlert } from '../../../components/ErrorAlert';

interface DataListPanelProps {
  onToggleFiltersPanel: () => void;
  previewItem: Record<string, unknown> | null;
  setPreviewItem: React.Dispatch<React.SetStateAction<Record<string, unknown> | null>>;
  listConfig?: ReturnType<typeof getListConfig>;
  cardFields?: ReturnType<typeof getCardFields>;
}

/**
 * Show a list of filterable `<DataListCard>` components based on the data source.
 * Cards are filterable by the inputs in `<FiltersPanel>` and clicking a card will
 * display the `<PreviewPanel>`.
 */
export const DataListPanel = ({
  onToggleFiltersPanel,
  previewItem,
  setPreviewItem,
  listConfig = getListConfig(),
  cardFields = getCardFields(),
}: DataListPanelProps) => {
  const { idField, title, content } = cardFields;
  const { source: dataSource } = listConfig;
  const [cards, setCards] = useState<Record<string, unknown>[]>([]);
  const [page, setPage] = useState(1);
  const [sortBy, setSortBy] = useState<'name' | 'date'>('name');

  const { data: response, isPending, isError, error } = useListQuery(dataSource);
  const data = response?.data || [];

  useEffect(() => {
    if (data.length > 0) {
      setCards(data);
    }
  }, [data]);

  const renderCards = useMemo(() => {
    if (isPending) {
      return <LoadingSpinner />;
    }
    if (isError) {
      return <ErrorAlert message={error?.message ?? 'Error fetching data'} />;
    }
    return cards.map((card, index) => {
      const cardId = String(card[idField as keyof typeof card] || index);
      const cardTitle = card[title as keyof typeof card] as string;
      const cardContent = card[content as keyof typeof card] as string;
      
      return (
        <DataListCard key={cardId} item={card} previewItem={previewItem} setPreviewItem={setPreviewItem}>
          <ListItemText
            primary={<Typography>{cardTitle}</Typography>}
            secondary={<Typography>{cardContent}</Typography>}
          />
        </DataListCard>
      );
    });
  }, [cards, isPending, isError, error, idField, title, content, previewItem, setPreviewItem]);

  // Content to render on the page for this component
  return (
    <Paper elevation={0}>
      <Box
        sx={{
          display: 'flex',
          flexDirection: 'column',
          height: '100%',
        }}
      >
        <Box
          sx={{
            display: 'flex',
            justifyContent: 'space-between',
            alignItems: 'center',
            padding: 2,
          }}
        >
          <Button
            startIcon={<FilterAltIcon />}
            onClick={onToggleFiltersPanel}
          >
            Filters
          </Button>
          <Button
            startIcon={<SortIcon />}
            onClick={() => setSortBy(sortBy === 'name' ? 'date' : 'name')}
          >
            Sort by {sortBy === 'name' ? 'Date' : 'Name'}
          </Button>
        </Box>
        <Box
          sx={{
            flex: 1,
            overflow: 'auto',
            padding: 2,
          }}
        >
          {renderCards}
          {cards && cards.length === 0 && (
            <Stack flex={1}>
              <Typography>No data matches your search</Typography>
            </Stack>
          )}
        </Box>
        <Box
          sx={{
            display: 'flex',
            justifyContent: 'center',
            padding: 2,
          }}
        >
          <Pagination
            count={Math.ceil((response?.total || 0) / 10)}
            page={page}
            onChange={(_, value) => setPage(value)}
          />
        </Box>
      </Box>
    </Paper>
  );
};
