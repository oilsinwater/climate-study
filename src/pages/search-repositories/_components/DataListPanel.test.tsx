import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import '@testing-library/jest-dom';
import { QueryClientProvider } from '@tanstack/react-query';
import { FilterContextProvider } from '../../../components/FilterContext';
import DataListPanel from './DataListPanel';
import { createTestQueryClient } from '../../../utils/testing.utils';
import { useListQuery } from '../../../utils/useListQuery';
import { vi } from 'vitest';

// Define mock data structure for testing
interface MockDataItem {
  id: string;
  name: string;
  date: string;
  description: string;
}

// Create mock data for testing
const mockData: MockDataItem[] = [
  {
    id: '1',
    name: 'Dataset A',
    date: '2023-01-01',
    description: 'Test dataset A',
  },
  {
    id: '2',
    name: 'Dataset B',
    date: '2023-02-01',
    description: 'Test dataset B',
  },
];

// Define complete Dataset interface matching the actual data structure
interface Dataset {
  id: string;
  title: string;
  summary: string;
  publication_date: string;
  citation: string;
  doi: string;
  source: string;
  tags: string[];
  category: string;
  point_of_contact: string;
  originator: string;
  metadata_contact: string;
  publisher: string;
  distributor: string;
  usgs_mission_area: string;
  communities: { title: string; url: string }[];
  associated_projects: { title: string; url: string }[];
  attached_files: {
    file_id: number;
    file_name: string;
    file_size: string;
    description: string;
  }[];
}

// Create a complete mock dataset for testing
const mockDataset = {
  id: 'test-id',
  title: 'Test Dataset',
  summary: 'Test summary',
  publication_date: '2023-01-01',
  citation: 'Test citation',
  doi: 'test-doi',
  source: 'Test source',
  tags: ['test'],
  category: 'test',
  point_of_contact: 'test@example.com',
  originator: 'Test Originator',
  metadata_contact: 'test@example.com',
  publisher: 'Test Publisher',
  distributor: 'Test Distributor',
  usgs_mission_area: 'Test Area',
  communities: [],
  associated_projects: [],
  attached_files: [],
};

// Add this above the test suite
const mockSearchResults: Dataset[] = [mockDataset];

// Mock the useListQuery hook for controlled testing
vi.mock('../../../utils/useListQuery', () => ({
  useListQuery: vi.fn(() => ({
    data: {
      data: mockData,
      total: mockData.length,
    },
    isPending: false,
    isError: false,
    error: null,
  })),
}));

// Main test suite for DataListPanel component
describe('DataListPanel', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  // Test loading state rendering
  it('renders loading state', async () => {
    // Mock loading state
    (useListQuery as vi.Mock).mockReturnValue({
      data: null,
      isPending: true,
      isError: false,
      error: null,
    });

    const mockSetPreviewItem = vi.fn();

    // Render component with test setup
    render(
      <QueryClientProvider client={createTestQueryClient()}>
        <FilterContextProvider>
          <DataListPanel
            searchResults={mockSearchResults}
            previewItem={null}
            setPreviewItem={mockSetPreviewItem}
          />
        </FilterContextProvider>
      </QueryClientProvider>
    );

    // Verify loading spinner is shown
    const spinner = await screen.findByRole('progressbar');
    expect(spinner).toBeInTheDocument();
  });

  // Test sorting functionality
  it('sorts by name ascending', async () => {
    // Mock successful data fetch
    (useListQuery as vi.Mock).mockReturnValue({
      data: mockData,
      isPending: false,
      isError: false,
      error: null,
    });

    const mockSetPreviewItem = vi.fn();

    // Render component
    render(
      <QueryClientProvider client={createTestQueryClient()}>
        <FilterContextProvider>
          <DataListPanel
            searchResults={mockSearchResults}
            previewItem={null}
            setPreviewItem={mockSetPreviewItem}
          />
        </FilterContextProvider>
      </QueryClientProvider>
    );

    // Test sorting interaction
    const sortButton = await screen.findByRole('button', {
      name: /sort by date/i,
    });
    fireEvent.click(sortButton);

    // Verify sorted order
    const items = await screen.findAllByRole('listitem');
    expect(items[0]).toHaveTextContent('Dataset A');
  });

  // Test error state handling
  it('displays error message when data fetching fails', async () => {
    // Mock error state
    vi.spyOn(console, 'error').mockImplementation(() => {});

    const mockSetPreviewItem = vi.fn();

    // Render component
    render(
      <QueryClientProvider client={createTestQueryClient()}>
        <FilterContextProvider>
          <DataListPanel
            searchResults={mockSearchResults}
            previewItem={null}
            setPreviewItem={mockSetPreviewItem}
          />
        </FilterContextProvider>
      </QueryClientProvider>
    );

    // Verify error message is shown
    const errorMessage = await screen.findByText(/error fetching data/i);
    expect(errorMessage).toBeInTheDocument();
  });

  // Test snapshot matching
  it('should render dataset items', () => {
    const mockPreviewItem = mockDataset;
    const mockSetPreviewItem = vi.fn();

    // Render component and verify snapshot
    const { container } = render(
      <QueryClientProvider client={createTestQueryClient()}>
        <FilterContextProvider>
          <DataListPanel
            searchResults={[mockDataset]}
            previewItem={mockPreviewItem}
            setPreviewItem={mockSetPreviewItem}
          />
        </FilterContextProvider>
      </QueryClientProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
