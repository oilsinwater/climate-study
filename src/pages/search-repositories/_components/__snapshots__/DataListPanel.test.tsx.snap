// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataListPanel > should render dataset items 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiStack-root css-gmwslw-MuiStack-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        Datasets
      </h6>
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
      >
        <ul
          class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateBeforeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-current="true"
              aria-label="page 1"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              tabindex="0"
              type="button"
            >
              1
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateNextIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
    <div
      class="MuiBox-root css-1yuhvjn"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1dq7s1d-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiBox-root css-4g6ai3"
        >
          <div
            class="MuiBox-root css-19midj6"
          >
            <div
              class="MuiStack-root css-1sos3zc-MuiStack-root"
            >
              <div
                class="MuiBox-root css-1rr4qq7"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                >
                  Test Dataset
                </h6>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-1xwhmis-MuiTypography-root"
                >
                  Test summary
                </p>
              </div>
              <button
                aria-label="view details"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
